### PR TITLE
[FLINK-21085][runtime][checkpoint] Allows triggering checkpoint for non-source tasks

### PR DESCRIPTION
**Pull request by gaoyunhaii**
## What is the purpose of the change

This PR enables triggering checkpoint barrier handle for non-source tasks. 

To support checkpoints after some tasks finished, non-source tasks might also receive RPC trigger of checkpoint after all their precedent tasks get finished. In this case, it need to trigger checkpoint barrier handler to insert barriers. This PR provides the functionality to notify checkpoint barrier on RPC trigger

## Brief change log

  - a3daa2b88a7c1609daf2bb55742971eabe5dacf0 refactors the current implementation to expose CheckpointBarrierHandler.
  - b4baae6e24556aec841fb1c4fcedfec234e3c0cd notifies the non-source stream tasks' checkpoint barrier handler on RPC trigger.
  - 326d8751ac2aac4255b5e94da8e8f46a424c55c5 fixes the tests that using non-source task that directly perform checkpoint on trigger.

## Verifying this change

This PR does not come with test right now since it does not have affect to the current process, and it is not easy to write tests since the CheckpointBarrierHandler is not updated yet. We would add the unit tests as a whole in the next PR, which also modifies the `CheckpointBarrierHandler`'s logic.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 1]
```diff
@@ -877,12 +875,6 @@ public StreamStatusMaintainer getStreamStatusMaintainer() {
         CompletableFuture<Boolean> result = new CompletableFuture<>();
         mainMailboxExecutor.execute(
                 () -> {
-                    latestAsyncCheckpointStartDelayNanos =
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:880]: Why is this moved to the "inner" method? I found it a bit confusing that the "non-async" method not determines the start of the async part.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointableOneInputStreamTask.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointMetricsBuilder;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.execution.Environment;
+
+/** A test stream task that also response to the checkpoint trigger requirement. */
+public class CheckpointableOneInputStreamTask<IN, OUT> extends OneInputStreamTask<IN, OUT> {
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointableOneInputStreamTask.java:27]: Why don't the tests work anymore with just the regular `OneInputStreamTask`? They also have an implemented `triggerCheckpoint()`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TestCheckpointBarrierHandler.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.streaming.runtime.io.checkpointing.CheckpointBarrierHandler;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A test {@link TestCheckpointBarrierHandler} that records the history of checkpoint triggering.
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/TestCheckpointBarrierHandler.java:34]:
```suggestion
 * A {@link CheckpointBarrierHandler} for testing that records the history of checkpoint triggering.
```

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointableOneInputStreamTask.java (hunk 2 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CheckpointableOneInputStreamTask.java:27]: This is due to the logic for this test OneInputStream is different from the formal one in that it would trigger a snapshot (like source stream task) instead of go through the CheckpointBarrierHandler. They are mainly used in some test cases related to taking checkpoint on TM side.

After some rethink I think for these test they would better to directly call `triggerCheckpointOnBarrier` to simulate the case that the alignment has finished. Thus I removed `CheckpointableOneInputStreamTask` and modifies the related tests to call `triggerCheckpointOnBarrier` directly.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 1 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:880]: This variable is only used for the true source tasks which do not attached with a `CheckpointBarrierHandler`, thus initially I think it could be moved to the base class for the source tasks would make it more clear. 

The variable mainly measures the delay between JM triggering checkpoint and (source) tasks start to processing this checkpoint (namely `checkpointStartDelayNanos`), since move it to be inside `triggerCheckpoint` would only differs in the time of one method call, thus I think it would not cause too large difference, and we could rename the variable to `latestACheckpointStartDelayNanos`. Do you think this option would also be ok?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 4]
```diff
@@ -901,54 +893,12 @@ public StreamStatusMaintainer getStreamStatusMaintainer() {
         return result;
     }
 
-    private boolean triggerCheckpoint(
+    protected boolean triggerCheckpoint(
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:896]: Currently not making it abstract is mainly because `MultipleInputStreamTask`  has override `triggerCheckpointAsync`, if we make ` triggerCheckpoint` abstract, then it has implemented an empty method which it does not use. If we want to also make `MultipleInputStreamTask` do not need to override `triggerCheckpointAsync` directly, we have to make the completable future a parameter to `triggerCheckpoint` and let the triggerCheckpoint to complete the future instead of hide it from `triggerCheckpoint`.

[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: I also commit a diff to show the changes if we make `triggerCheckpoint` to be abstract: https://github.com/apache/flink/pull/14820/commits/83de6dda8fab862126c1e91010141828b93fe466

**Comment by dawidwys** [comment]
Honestly, I am not convinced about the changes. I don't like exposing the `CheckpointBarrierHandler`. Moreover the `triggerCheckpoint` method does not fit well into the class imo.

You already have two separate methods in the `StreamTask` for triggering a checkpoint either from RPC or on a barrier. I don't necessarily understand why does it have to do so many rounds through the `StreamTask#triggerCheckpointAsync` -> `CheckpointBarrierHandler#triggerCheckpoint` -> `StreamTask#triggerCheckpointOnBarrier` -> `StreamTask#performCheckpoint`. Can't you just remove the two middle man? I think a check if all inputs are finished in the `StreamTask#triggerCheckpointAsync` of non-source tasks should be enough.

**Comment by gaoyunhaii** [comment]
Hi @dawidwys very thanks for the suggestions!

The main blocker from my view for directly performing the checkpoint  is to subsuming or waiting for the previous checkpoints. For example, suppose we have

```
A ---|
     |--> C
B ---|
```

We may met the cases like

1. A emits Barrier 5.
2. A finished and emits EndOfPartition, but C has not received yet.
3. B finished and emits EndOfPartition, but C has not received yet.
4. The CheckpointCoordinator notifies C for checkpoint 6.

In this case we either wait for checkpoint 5 finish and then trigger checkpoint 6, or subsuming the checkpoint 5 and trigger checkpoint 6. But in both cases, only `CheckpointBarrierHandler` knows the existence of checkpoint 5, thus it seems we have to notify the `CheckpointBarrierHandler` in some way ? 

One possible alternative might be let the `CheckpointBarrierHandler` first do a check of outdatedness on alignment, if so, then abort the checkpoint. The drawback of this method is that the `CheckpointBarrierHandler` might do some useless work, and we would have two entrances for performCheckpoint() that are not fully independent.

**Comment by dawidwys** [comment]
I see. Let me ask an additional question. The problem applies only for UC, right? For AC that situation is not possible, because aligned barriers do not overtake the data buffers and tasks won't finish before the final aligned checkpoint. Would it work if we force such barriers to be aligned? I am trying to understand all the options here.

**Comment by gaoyunhaii** [comment]
Hi @dawidwys I think this might be still happen in AC since 

1. For the final version not all tasks might need to wait for the last aligned checkpoint since some tasks do not have operators using 2pc. 
2. There might be still some cases even if `A` and `B` all need to wait for the last checkpoints. Suppose they are all waiting for checkpoint 4, with concurrent checkpoints we might start triggering checkpoint 5. However the network stalled after emitting barrier 4. Then we finally have checkpoint 4 completed, notify `A` and `B` to make them emit `EndOfPartition` and exit. Then if JM trigger checkpoint 6 now we would have the similar case.

**Comment by dawidwys** [comment]
All right, I will need to think about it a bit more, but even assuming we can receive an RPC signal while the input channels are still active/non-finished I feel we can make it easier. I don't think there is a big enough difference between source and non-source tasks to justify extracting a separate hierarchy. The way I see it is that the only difference is if all the input gates has finished or not. (If there are no input gates all of them are obviously finished ;) ). In that scenario we could extend the `triggerCheckpointAsyncInMailbox` (it was renamed on master) to handle both cases quite easily.

I think we don't need to expose the `SingleCheckpointBarrierHandler` in any way. Either with my approach or yours you would need to inject barriers on the RPC call into the channels/or drain the channels somehow, right? Otherwise you cannot checkpoint the data/align with the data in the channels properly. Or am I missing something? Therefore you could inject the barrier into channel/drain the channel from the `StreamTask` level and the logic in `SingleCheckpointBarrierHandler` would still be valid.

Let me know what you think.

```
    @Override
    public Future<Boolean> triggerCheckpointAsync(
            CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {

        CompletableFuture<Boolean> result = new CompletableFuture<>();
        mainMailboxExecutor.execute(
                () -> {
                    try {
                        result.complete(
                                triggerCheckpointAsyncInMailbox(
                                        checkpointMetaData, checkpointOptions));
                    } catch (Exception ex) {
                        // Report the failure both via the Future result but also to the mailbox
                        result.completeExceptionally(ex);
                        throw ex;
                    }
                },
                "checkpoint %s with %s",
                checkpointMetaData,
                checkpointOptions);
        return result;
    }

    private boolean triggerCheckpointAsyncInMailbox(
            CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions)
            throws Exception {
        FlinkSecurityManager.monitorUserSystemExitForCurrentThread();
        try {
            if (allInputsFinished()) {
                return triggerCheckpointInRootNode(checkpointMetaData, checkpointOptions);
            } else {
                throw new UnsupportedOperationException(
                        "We do not support triggering non root nodes yet.");
            }
        } catch (Exception e) {
            // propagate exceptions only if the task is still in "running" state
            if (isRunning) {
                throw new Exception(
                        "Could not perform checkpoint "
                                + checkpointMetaData.getCheckpointId()
                                + " for operator "
                                + getName()
                                + '.',
                        e);
            } else {
                LOG.debug(
                        "Could not perform checkpoint {} for operator {} while the "
                                + "invokable was not in state running.",
                        checkpointMetaData.getCheckpointId(),
                        getName(),
                        e);
                return false;
            }
        } finally {
            FlinkSecurityManager.unmonitorUserSystemExitForCurrentThread();
        }
    }

    private boolean allInputsFinished() {
        return getEnvironment().getAllInputGates().length == 0
                || Arrays.stream(getEnvironment().getAllInputGates())
                        .allMatch(InputGate::isFinished);
    }

    private boolean triggerCheckpointInRootNode(
            CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions)
            throws Exception {
        latestAsyncCheckpointStartDelayNanos =
                1_000_000
                        * Math.max(
                                0, System.currentTimeMillis() - checkpointMetaData.getTimestamp());

        // No alignment if we inject a checkpoint
        CheckpointMetricsBuilder checkpointMetrics =
                new CheckpointMetricsBuilder()
                        .setAlignmentDurationNanos(0L)
                        .setBytesProcessedDuringAlignment(0L)
                        .setCheckpointStartDelayNanos(latestAsyncCheckpointStartDelayNanos);

        subtaskCheckpointCoordinator.initInputsCheckpoint(
                checkpointMetaData.getCheckpointId(), checkpointOptions);

        boolean success =
                performCheckpoint(checkpointMetaData, checkpointOptions, checkpointMetrics);
        if (!success) {
            declineCheckpoint(checkpointMetaData.getCheckpointId());
        }
        return success;
    }
```

**Comment by gaoyunhaii** [comment]
Hi Dawid, very thanks for the suggestion! After some more think, now I think you are right that we would finally drain the records from all the channels and we might handle the triggering the remaining checkpoints directly in the `StreamTask` so that we do not need to expose the `BarrierHandler`. 

With the above approach, it seems we would reject the checkpoints if not all channels get finished, right ? Perhaps we could change it a bit to holding these checkpoints until all the checkpoints get drained ? namely

```
private final List<CheckpointBarrier> pendingCheckpoints = new ArrayList();

private boolean triggerCheckpointAsyncInMailbox(
        CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions)
        throws Exception {
    FlinkSecurityManager.monitorUserSystemExitForCurrentThread();
    try {
        if (allInputsFinished()) {
            return triggerCheckpointInRootNode(checkpointMetaData, checkpointOptions);
        } else {
            pendingCheckpoints.add(new CheckpointBarrier(checkpointMetaData, checkpointOptions));
        }
    } catch (Exception e) {
       ...
    } finally {
        ...
    }
}

private void executeInvoke() throws Exception {
    runMailboxLoop();
    ensureNotCanceled();

    triggerAllPendingCheckpoints();

    afterInvoke();
}


private void triggerAllPendingCheckpoints() {
    for (CheckpointBarrier barrier : pendingCheckpoints) {
         triggerCheckpointInRootNode(barrier.getMetaData(), barrier.getOptions());
    }
   pendingCheckpoints.clear();
}

```

Do you think this would be ok~?

**Comment by dawidwys** [comment]
I thought about it quite a lot yesterday. Sorry for going back and forth with the issue. However, I want to be extra cautious with the checkpointing. I checked the FLIP again and also discussed with Piotr what would be the best general approach for handling the triggering in those corner cases.

Let me, recollect what we said in the FLIP. IIRC we will introduce an additional `EndOfData` event which will be hand-shaken between the upstream and downstream task before the upstream task goes away, right? This means it is not possible to get an RPC `triggerCheckpoint` on a downstream task if there is any data in any of the channels. 

Still, there might be barriers incoming through channels that already reported the `EndOfData` as the upstream tasks might've not gone away. In the discussion with Piotr, we said it would be better to minimise the number of declined/delayed checkpoints. Having said that I think your original proposal could come in handy here. 

I'd still do the logic within the `StreamTask`, but I think it's a good idea to make the `CheckpointBarrierHandler` available here. Then we could handle the RPC in the following manner:

```
        List<?> notFinishedInputs = getNotFinishedInputs() // just a pseud code, we could probably do it cleaner
        if (notFinishedInputs.size() == 0) { // channels already closed, we won't see any data nor barriers
            return triggerCheckpointInRootNode(checkpointMetaData, checkpointOptions);
        } else { // we have seen the EndOfData, but there might be some barriers arriving

            // theoretically we could trigger/processBarrier for just a single channel, as all channels should've seen the EndOfData and they should not participate in the alignment, we care only about the triggering, however for the sake of completeness I'd do it for all the not finished channels

            for (InputGate inputGate : notFinishedInputs) {
                for (InputChannelInfo channelInfo : inputGate.getNotFinishedChannels()) {
                     getCheckpointBarrierHandler().processBarrier(barrier, channelInfo);
                }
            }
        }
```

Therefore it's very much like your original proposal, just that we do not introduce the additional Source/Non-source split and we do not need to expose `isRunning` or `performCheckpoint`.

WDYT?

**Comment by gaoyunhaii** [comment]
Hi Dawid, very thanks for the thoughts! 

I think the analysis on the `EndOfDataEvent` is exactly right, currently when a task get triggered, all the user records must be processed and there are only possible barriers remaining. 

For the current approach, I think it avoids the source / non-source split since it unified the logic of triggering checkpoints by distinguish the different situation via `getNotFinishedChannels()`, right ? And `getCheckpointBarrierHandler()` seems need to be an abstract method that source return null while non-source returns the actual `CheckpointBarrier`? 

I think I am also agree with this method. I'll update the PR according to this approach. very thanks for the suggestions!

**Comment by gaoyunhaii** [comment]
Hi @dawidwys, I updated the PR according the above discussion. Some points are met during the discussion that might need consideration:

1. For `MultipleInputStreamTask`, if there is no chained source, it would behave like ordinary `StreamTask`. But if there are sources chained, the task would always get triggered since it is also a kind of source, and we would always only trigger the chained sources to initiate the checkpoint on the task side. Then if the checkpoint is triggered between the upstream tasks are finished and `MultipleInputStreamTask` received the `EndOfPartitionEvent`, we could not subsume the pending checkpoints like the other `StreamTask` do. But since in this case the only possible records are `EndOfPartitionEvent` and `CheckpointBarrier`, logically they would not be delayed too long, and it only have affect in a short of period, I think it might be acceptable. Thus do you think this would be a problem ? 
2. Regarding how to acquire the unfinished channels from `IndexInputGate`, returning an `Iterable<InputChannelInfo>` or pass a `Consumer<InputChannelInfo>` could avoid allocating additional memory, but they all not work well in considering that we need to access `channelsWithEndOfPartitionEvents` in the lock scope. Thus I think we might still need to return a copied list of unfinished input channels.

Very thanks for the review!

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Comment by dawidwys** [comment]
Ad. 1 I'd say we should differentiate two cases. If it does not have any chained sources it should behave the same way as the `StreamTask`. If we have chained sources, I think there is no problem, because triggering a new checkpoint will be done anyway for the sources. For the remaining non-source inputs the `EndOfPartitionEvent` will act as a checkpoint barrier basically. 

Which is basically what you're saying as well as far as I can tell ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 5]
```diff
@@ -1012,6 +1027,35 @@ private boolean triggerCheckpointAsyncInMailbox(
...
+                for (InputChannelInfo channelInfo : inputGate.getUnfinishedChannels()) {
+                    checkpointBarrierHandler.get().processBarrier(barrier, channelInfo);
+                }
+            }
+        }
+
+        return true;
+    }
+
+    protected Optional<CheckpointBarrierHandler> getCheckpointBarrierHandler() {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:1055]: Could you add a javadoc here? I think it will be helpful to explain when can we return `Optional.empty()` here and what are the consequences. That there should be no inputs in that case.

**Review by dawidwys — COMMENTED** [review]
I need to go over the tests (I'll do that tomorrow), but it looks rather good. I would appreciate adding a javadoc. I will most probably merge it after going through the tests.

**Comment by dawidwys** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 5 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:1060]: I updated the javadoc~

**Comment by gaoyunhaii** [comment]
Hi @dawidwys  very thanks for the review and glad that we think similarly on the first point ;-)  I have updated the PR according to the comments~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java [hunk 6]
```diff
@@ -175,10 +176,23 @@ protected void createInputProcessor(
...
 
+        if (operatorChain.getSourceTaskInputs().size() == 0) {
+            return super.triggerCheckpointAsync(metadata, options);
+        }
+
+        // If there are chained sources, we would always only trigger
+        // the chained sources for checkpoint. This means that for
+        // the checkpoints during the upstream task finished and
+        // this task receives the EndOfPartitionEvent, the checkpoint
+        // would not subsume the pending ones.
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java:195]: Is that comment correct? We will still subsume the pending checkpoints from within the sources. We do process barriers for the source inputs in that case.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java [hunk 7]
```diff
@@ -77,6 +77,12 @@
 
     private static final int CHECKPOINT_EXECUTION_DELAY_LOG_THRESHOLD_MS = 30_000;
 
+    /**
+     * TODO Whether enables checkpoints after tasks finished. This is a temporary flag and will be
+     * removed in the last PR.
+     */
+    protected boolean enableCheckpointAfterTasksFinished;
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:84]: Why is this `protected`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java [hunk 8]
```diff
@@ -1723,6 +1725,86 @@ protected void cancelTask() {
         assertTrue(OpenFailingOperator.wasClosed);
     }
 
+    @Test
+    public void testTriggeringCheckpointWithFinishedChannels() throws Exception {
+        OneInputStreamTaskTestHarness<String, String> testHarness =
```
Comment (dawidwys) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1730]: Could you rewrite the test to the newer `StreamTaskMailboxTestHarness`. I believe it should be fairly straightforward.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java (hunk 6 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java:195]: Yes, we should indeed also be able to subsume the checkpoints in this case, and the actual difference is that we would not insert barriers for the unfinished network inputs and it needs to wait for the EndOfPartitionEvent. I'll fix the comments~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java (hunk 7 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java]: This should be a mistake... I changed it to private

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java (hunk 8 above)
Comment (gaoyunhaii) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java]: I updated the test~

**Comment by gaoyunhaii** [comment]
Hi @dawidwys  very thanks for the review~ I updated the PR according to the comments~
